apps/solaredgemonitor/solaredge_monitor.star

```diff
@@ -5,167 +5,671 @@ Description: Energy production and consumption monitor for your SolarEdge solar
+                expanded = True,
                 cross_align = "center",
                 children = [
-                    render.Image(src = img),
-                    render.Text(
-                        content = format_power(power),
-                        height = 8,
-                        font = "tb-8",
-                        color = "#ffd11a",
+                    render.Row(
+                        expanded = True,
+                        main_align = "space_evenly",
+                        cross_align = "center",
+                        children = [
+                            render.Text("Energy Today"),
+                        ],
                     ),
-                    render.Text(
-                        content = unit,
-                        height = 8,
-                        font = "tb-8",
-                        color = "#ffd11a",
+                    render.Row(
+                        #expanded = True,
+                        children = [
+                            render.Column(
+                                expanded = True,
+                                main_align = "space_around",
+                                cross_align = "center",
+                                children = [
+                                    render.Image(src = SUN_SUM),
+                                    render.Image(src = PLUG_SUM),
+                                ],
+                            ),
+                            render.Column(
+                                expanded = True,
+                                main_align = "space_around",
+                                cross_align = "end",
+                                children = [
+                                    render.Text(
+                                        content = " " + format_power(production / 1000) + " kWh",
+                                        font = "5x8",
+                                        color = GREEN,
+                                    ),
+                                    render.Text(
+                                        content = " " + format_power(consumption / 1000) + " kWh",
+                                        font = "5x8",
+                                        color = RED,
+                                    ),
+                                ],
+                            ),
+                        ],
                     ),
                 ],
             ),
-        )
-    dots = [render.Box(width = 18)]
-    for dir in flows:
-        if dir:
-            el = render.Image(src = DOTS_LTR if dir == 1 else DOTS_RTL)
-        else:
-            el = render.Box(width = 10)
-        dots.append(
-            render.Stack(children = [render.Box(width = 21), el]),
-        )
-    return render.Root(
-        child = render.Stack(
+        ],
+    )
+
+    # CHARGE FRAME shows charge/discharge rate and state of charge percent
```

Comment:
protect this entire block including charge_frame with `if has_battery` to avoid crash on `o["STORAGE"]`

Reply:
Done. I've applied this to all frames so that each is only generated if needed.